src/compiler/JSCodeGenVisitor.js

```diff
@@ -42,16 +47,31 @@ module.exports = class JSCodeGenVitor {
     this._indent = this._indents[++this._indentLevel] || ''; // eslint-disable-line no-plusplus
     return this;
   }
+
   outdent() {
     this._indent = this._indents[--this._indentLevel] || ''; // eslint-disable-line no-plusplus
     return this;
   }
 
+  setIndent(n) {
+    this._indentLevel = n;
+    this._indent = this._indents[n] || '';
+    return this;
+  }
+
   process(commands) {
+    const definedVariables = [];
+
     // first define the globals so they are accessible from any block below
     commands.forEach((c) => {
       if (c instanceof VariableBinding.Global) {
         this._out(`let ${c.variableName};`);
```

Comment:
There is an edge-case here; this will throw (redeclaring a variable name) if the user defines a global, like `foo` that also matches a template name.

In this case, I'm not sure what should take precedence; should we prevent template names from being output if a global is already defined?

Reply:
I think how the template name is handled is implementation detail. using it directly as function name is probably not good. similar the `call` plugin must validate that the template really exists and not just call any function (like `eval`). how about auto-generated prefixes and a mapping table?